src/meta/processors/admin/Balancer.cpp

```diff
@@ -0,0 +1,264 @@
+                            std::inserter(diff, diff.begin()));
+        for (auto& partId : diff) {
+            if (partsFrom.size() >= avgLoad && partsTo.size() <= avgLoad) {
+                auto it = std::find(partsFrom.begin(), partsFrom.end(), partId);
+                CHECK(it != partsFrom.end());
+                partsFrom.erase(it);
+                partsTo.emplace_back(partId);
+                tasks.emplace_back(partId, maxPartsHost.first, minPartsHost.first, kv_);
+            }
+        }
+        hosts = sortedHostsByParts(newHostParts);
+        maxPartsHost = hosts.back();
+        minPartsHost = hosts.front();
+    }
+    LOG(INFO) << "I am balanced now!";
+}
+
+void Balancer::getHostParts(GraphSpaceID spaceId,
+                            std::unordered_map<HostAddr, std::vector<PartitionID>>& hostParts,
+                            int32_t& totalParts) {
+    folly::SharedMutex::ReadHolder rHolder(LockUtils::spaceLock());
+    auto prefix = MetaServiceUtils::partPrefix(spaceId);
+    std::unique_ptr<kvstore::KVIterator> iter;
+    auto ret = kv_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+    if (ret != kvstore::ResultCode::SUCCEEDED) {
+        LOG(ERROR) << "Access kvstore failed, spaceId " << spaceId;
+        return;
+    }
+    while (iter->valid()) {
+        auto key = iter->key();
+        PartitionID partId;
+        memcpy(&partId, key.data() + prefix.size(), sizeof(PartitionID));
+        auto partHosts = MetaServiceUtils::parsePartVal(iter->val());
+        for (auto& ph : partHosts) {
+            hostParts[HostAddr(ph.ip, ph.port)].emplace_back(partId);
+        }
+        iter->next();
+        totalParts++;
+    }
+}
+
+void Balancer::calDiff(const std::unordered_map<HostAddr, std::vector<PartitionID>>& hostParts,
+                       const std::vector<HostAddr>& activeHosts,
+                       std::vector<HostAddr>& newlyAdded,
+                       std::vector<HostAddr>& lost) {
+    for (auto it = hostParts.begin(); it != hostParts.end(); it++) {
+        if (std::find(activeHosts.begin(), activeHosts.end(), it->first) == activeHosts.end()) {
+            lost.emplace_back(it->first);
+        }
+    }
+    for (auto& h : activeHosts) {
+        if (hostParts.find(h) == hostParts.end()) {
+            newlyAdded.emplace_back(h);
+        }
+    }
+}
+
+std::vector<std::pair<HostAddr, int32_t>>
+Balancer::sortedHostsByParts(const std::unordered_map<HostAddr,
+                                                      std::vector<PartitionID>>& hostParts) {
+    std::vector<std::pair<HostAddr, int32_t>> hosts;
+    for (auto it = hostParts.begin(); it != hostParts.end(); it++) {
+        hosts.emplace_back(it->first, it->second.size());
+    }
+    std::sort(hosts.begin(), hosts.end(), [](const auto& l, const auto& r) {
+        return l.second < r.second;
+    });
+    return hosts;
+}
+
+StatusOr<HostAddr> Balancer::pickupHost(
```

Comment:
May be we should give the function with a more meaningful name: `hostWithMinimalParts`, WDYT?

Reply:
IMO pickupHost could be better to be a interface function. what if another pickup method (by working load?).

Reply:
HA...  I think both of them are OK now.  
As @spacewalkman mentioned above,  hostWithMinimalParts could be more direct.  Maybe we could change it with another name when the algorithm changed.